### PR TITLE
mdc auto injection integration

### PR DESCRIPTION
**Pull request by realark**
* Add post-match hook to default instrumenter api
* Add internal scope listener api
* slf4j MDC autoinjection integration

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java [hunk 1]
```diff
@@ -92,7 +92,6 @@ public static ResettableClassFileTransformer installBytebuddyAgent(
             .or(nameStartsWith("org.aspectj."))
             .or(nameStartsWith("org.groovy."))
             .or(nameStartsWith("com.p6spy."))
-            .or(nameStartsWith("org.slf4j."))
```
Comment (realark) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java]: This was intentional. This is actually not doing anything currently because shadow will rename the prefix to `datadog.slf4j`, which is already skipped.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java [hunk 2]
```diff
@@ -86,6 +86,24 @@ public final AgentBuilder instrument(final AgentBuilder parentAgentBuilder) {
...
+                        ProtectionDomain protectionDomain) {
+                      postMatch(
+                          typeDescription,
+                          classLoader,
+                          module,
+                          classBeingRedefined,
+                          protectionDomain);
+                      return true;
+                    }
+                  })
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java]: Would it make sense to pull this out into an actual named class like we do for MuzzleMatcher above to make this method more readable?

File: gradle/java.gradle [hunk 3]
```diff
@@ -15,12 +15,15 @@ lombok { // optional: values below are the defaults
   sha256 = ""
 }
 
-apply plugin: "net.ltgt.errorprone"
-tasks.withType(JavaCompile) {
-  options.compilerArgs += ['-Xep:FutureReturnValueIgnored:OFF']
-  // workaround for: https://github.com/google/error-prone/issues/780
-  options.compilerArgs += ['-Xep:ParameterName:OFF']
-  options.compilerArgs += ['-XepDisableWarningsInGeneratedCode']
+if (project.hasProperty('enableErrorProne')) {
```
Comment (mar-kolya) [inline gradle/java.gradle]: Could you please explain what is the reason for this change?

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  // mdcClassName = org.slf4j.MDC
+  private static final String mdcClassName = "org.TMP.MDC".replaceFirst("TMP", "slf4j");
+
+  public MDCInjectionInstrumentation() {
+    super(MDC_INSTRUMENTATION_NAME);
+  }
+
+  @Override
+  protected boolean defaultEnabled() {
+    final String enableInjection = getPropOrEnv("dd." + Config.LOGS_INJECTION);
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: If `Config` getter not available here?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java [hunk 5]
```diff
@@ -179,6 +197,23 @@ protected ReferenceMatcher getInstrumentationMuzzle() {
     /** @return A type matcher used to match the class under transform. */
     public abstract ElementMatcher<? super TypeDescription> typeMatcher();
 
+    /**
+     * A hook invoked after matching has succeeded and before transformers have run.
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java]: Could you please add some sort of a note here that this is being run from matcher context, not from transformation context?

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java [hunk 6]
```diff
@@ -91,6 +94,8 @@
   @Getter private final Integer jmxFetchRefreshBeansPeriod;
   @Getter private final String jmxFetchStatsdHost;
   @Getter private final Integer jmxFetchStatsdPort;
+  // Use (capital) Boolean to prefix generated getter with "get"
+  @Getter private final Boolean logsInjection;
```
Comment (mar-kolya) [inline dd-trace-api/src/main/java/datadog/trace/api/Config.java]: you may want to make this `boolean logsInjectionEnabled` to avoid boxed fields.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java (hunk 2 above)
Comment (realark) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java]: 👍

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  @Override
+  public void postMatch(
+      TypeDescription typeDescription,
+      ClassLoader classLoader,
+      JavaModule module,
+      Class<?> classBeingRedefined,
+      ProtectionDomain protectionDomain) {
+    MDCAdvice.mdcClassInitialized(classBeingRedefined);
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: This method is annotated with BB stuff, are we sure it is safe to call is from 'outside' of BB logic?

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  public static class MDCAdvice {
+    @Advice.OnMethodExit
+    public static void mdcClassInitialized(@Advice.Origin final Class mdcClass) {
+      try {
+        Method putMethod;
+        try {
+          putMethod = mdcClass.getMethod("put", String.class, String.class);
+        } catch (NoSuchMethodException e) {
+          putMethod = mdcClass.getMethod("put", String.class, Object.class);
+        }
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: Could you please add a comment about why there is this try/catch and why you do not need one for remove.

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        isTypeInitializer(), MDCAdvice.class.getName());
+  }
+
+  @Override
+  public String[] helperClassNames() {
+    return new String[] {MDCAdvice.class.getName() + "$MDCScopeListener"};
+  }
+
+  public static class MDCAdvice {
+    @Advice.OnMethodExit
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: Should there be 'suppress' stuff?

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java (hunk 6 above)
Comment (realark) [reply dd-trace-api/src/main/java/datadog/trace/api/Config.java]: `isLogsInjection` sounds awkward, but I guess I'm okay with it.

File: gradle/java.gradle (hunk 3 above)
Comment (realark) [reply gradle/java.gradle]: I left an explanation in the top-level PR body.

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        GlobalTracer.get().addScopeListener(new MDCScopeListener(putMethod, removeMethod));
+      } catch (NoSuchMethodException e) {
+        org.slf4j.LoggerFactory.getLogger(mdcClass).debug("Failed to add MDC span listener", e);
+      }
+    }
+
+    public static class MDCScopeListener implements ScopeListener {
+      private static final Logger log = org.slf4j.LoggerFactory.getLogger("mdc-scope-listener");
+      private static final String traceIdKey = "dd.trace_id";
+      private static final String spanIdKey = "dd.span_id";
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: These string seem to be constants, could you please capitalize them. Also you may want to make them visible for tests to avoid repeating them there.

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java (hunk 4 above)
Comment (realark) [reply dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: It is available, but our instrumentation seems to use a parallel path to get the configuration. I tried originally using `Config.get()`, but that isn't compatible with our testing.

That can be addressed but it's more of a general instrumenter thing, so I'd rather not fix it in this PR.

File: dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+   * Called just after a scope becomes the active scope
+   *
+   * <p>May be called multiple times. When a scope is initially created, or after a child scope is
+   * deactivated.
+   */
+  void afterScopeActivated();
+
+  /** Called just after a scope is closed. */
+  void afterScopeClose();
+}
```
Comment (mar-kolya) [inline dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java]: This seems to be highly asymmetrical API. I would suggest changing this to:

```
void afterScopeActivated();

void beforeScopeDeactivated();
```

And make sure that these are called whenever active scope changes (i.e. `beforeScopeDeactivated` is called on a scope being replaced as well).

File: dd-trace-ot/src/test/groovy/datadog/opentracing/scopemanager/ScopeManagerTest.groovy [hunk 12]
```diff
@@ -407,6 +409,52 @@ class ScopeManagerTest extends Specification {
     [new AtomicReferenceScope(true), new AtomicReferenceScope(true)] | _
   }
 
+  def "add scope listener"() {
```
Comment (mar-kolya) [inline dd-trace-ot/src/test/groovy/datadog/opentracing/scopemanager/ScopeManagerTest.groovy]: Looks like changes to `ContinuableScope` remain untested.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java [hunk 13]
```diff
@@ -86,6 +86,24 @@ public final AgentBuilder instrument(final AgentBuilder parentAgentBuilder) {
                       "Instrumentation class loader matcher unexpected exception: "
                           + getClass().getName()))
               .and(new MuzzleMatcher())
+              .and(
+                  new AgentBuilder.RawMatcher() {
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java]: Would it be possible/better to only add this if a postMatch is defined?  (maybe `!(this instanceof Default)`?)

File: dd-trace-ot/src/test/groovy/datadog/opentracing/scopemanager/ScopeManagerTest.groovy (hunk 12 above)
Comment (mar-kolya) [reply dd-trace-ot/src/test/groovy/datadog/opentracing/scopemanager/ScopeManagerTest.groovy]: I see `ContinuableScope` changes are exercised by `Slf4jMDCTest` but it still would be nice to have tests for them 'locally'.

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java (hunk 10 above)
Comment (tylerbenson) [inline dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: Are these keys defined in a spec? Do we want to allow users to configure different keys?

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        }
+        final Method removeMethod = mdcClass.getMethod("remove", String.class);
+        GlobalTracer.get().addScopeListener(new MDCScopeListener(putMethod, removeMethod));
+      } catch (NoSuchMethodException e) {
+        org.slf4j.LoggerFactory.getLogger(mdcClass).debug("Failed to add MDC span listener", e);
+      }
+    }
+
+    public static class MDCScopeListener implements ScopeListener {
+      private static final Logger log = org.slf4j.LoggerFactory.getLogger("mdc-scope-listener");
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: You don't want to use `@Slf4j`?

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java (hunk 8 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: Maybe include the versions on which the signature changed?

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java [hunk 15]
```diff
@@ -59,6 +59,9 @@
   public static final String DEFAULT_AGENT_HOST = "localhost";
   public static final int DEFAULT_TRACE_AGENT_PORT = 8126;
 
+  public static final String LOGS_INJECTION = "logs.injection";
```
Comment (tylerbenson) [inline dd-trace-api/src/main/java/datadog/trace/api/Config.java]: This is fine, but throwing out some alternatives:
* `logs.correlation`
* `logs.associate`

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java (hunk 6 above)
Comment (tylerbenson) [reply dd-trace-api/src/main/java/datadog/trace/api/Config.java]: The annotation might allow customizing the method name if you want something different.

File: dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
+package datadog.trace.context;
+
+/** Hooks for scope activation */
+public interface ScopeListener {
```
Comment (tylerbenson) [inline dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java]: Are we sure we want this on the public API right now?

File: dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+  /**
+   * Called just after a scope becomes the active scope
+   *
+   * <p>May be called multiple times. When a scope is initially created, or after a child scope is
+   * deactivated.
+   */
+  void afterScopeActivated();
```
Comment (tylerbenson) [inline dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java]: Would it be better to allow the scope/span to be passed into both methods?  It might not be necessary here, but I can imagine it might be useful in other cases.

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java (hunk 6 above)
Comment (mar-kolya) [reply dd-trace-api/src/main/java/datadog/trace/api/Config.java]: `logsInjectionEnabled` would lead to `isLogsInjectionEnabled` which sounds reasonable to me :)

File: gradle/java.gradle (hunk 3 above)
Comment (mar-kolya) [reply gradle/java.gradle]: Right. I missed that.... So two questions here:
* Do we still need to keep this code. It seems like deadwood at this point and can be removed
* If we are planning to keep it maybe we should have that comment from top of PR in the code to at least know why that is disabled.

Comment (realark) [reply gradle/java.gradle]: Good point. I think I'll remove it entirely. I don't think we get much value out of errorprone.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java (hunk 5 above)
Comment (realark) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java]: 👍

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java (hunk 6 above)
Comment (realark) [reply dd-trace-api/src/main/java/datadog/trace/api/Config.java]: The annotation doesn't allow customizing the name in this case unfortunately.

`logsInjectionEnabled` sounds good!

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java (hunk 7 above)
Comment (realark) [reply dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: It's safe to do in this case specifically because the advice only uses bootstrap classes. Also, this bit of code runs underneath the jvm transformer, so any errors can't escape into the application.

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java (hunk 8 above)
Comment (realark) [reply dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: That was actually something I was doing for logback specifically. I ended up only instrumenting the slf4j api, so I'll remove the `Object` fallback.

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java (hunk 9 above)
Comment (realark) [reply dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: Yes, thanks!

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java (hunk 10 above)
Comment (realark) [reply dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: 👍

File: dd-trace-ot/src/test/groovy/datadog/opentracing/scopemanager/ScopeManagerTest.groovy (hunk 12 above)
Comment (realark) [reply dd-trace-ot/src/test/groovy/datadog/opentracing/scopemanager/ScopeManagerTest.groovy]: Sure, I think that's easy to add.

File: dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java (hunk 14 above)
Comment (realark) [reply dd-java-agent/instrumentation/slf4j-mdc/src/main/java/datadog/trace/instrumentation/slf4j/mdc/MDCInjectionInstrumentation.java]: 👍 

I'm in the habit of not doing that in instrumentation, but it would work in this case since it's a helper class.

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java (hunk 15 above)
Comment (realark) [reply dd-trace-api/src/main/java/datadog/trace/api/Config.java]: I think those are fine, but we'll need an rfc change first. We're just following the rfc for configuring a common env variable.

File: dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java (hunk 16 above)
Comment (realark) [reply dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java]: I'd rather not, but unfortunately this is the only project available to both dd-trace-ot and instrumentation.

We already mix semi-private APIs in this project (see other classes in context package). It's not great, but it's the lesser of two evils. The alternative would be to expose `agent-bootstrap` to `dd-trace-ot`, which then brings a lot of stuff into the published `dd-trace-ot` artifact.

File: dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java (hunk 17 above)
Comment (realark) [reply dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java]: I would like to do that, but `dd-trace-api` doesn't depend on OpenTracing. So we'd have create another scope-like interface or overload the existing TraceScope.

File: dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java (hunk 11 above)
Comment (mar-kolya) [reply dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java]: That's fine by be, but I guess in your terms `activated`/`deactivated` is essentially same things I've proposed with `afterScopeActivated`/`beforeScopeDeactivated`. I'm slightly concerned that this is declared public and is not in a state we want that to be longer term - at the same time.

File: gradle/java.gradle (hunk 3 above)
Comment (realark) [reply gradle/java.gradle]: Moving to https://github.com/DataDog/dd-trace-java/pull/632

File: dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java (hunk 11 above)
Comment (realark) [reply dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java]: Yes, they are the the same thing. I think your names would be better for a real api; I just meant to point out that the current ScopeListener is intentionally incomplete.

I have concerns about this as well, but it's the least evil option I see until we can replace `dd-trace-ot` with a the new api.

**Comment by realark** [comment]
@mar-kolya new changes pushed. I believe this address all the feedback Can you please re-review? Once the review gets to a good state I'll doe one final rebase for the `fixup:` commits.

File: dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java (hunk 11 above)
Comment (tylerbenson) [reply dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java]: Perhaps at least change it to `afterScopeClosed` to maintain some consistency.

Comment (realark) [reply dd-trace-api/src/main/java/datadog/trace/context/ScopeListener.java]: Works for me. Done
